integration-testing/src/test/java/io/grpc/testing/integration/NettyTransportIoEfficiencyTest.java

```diff
@@ -0,0 +1,344 @@
+            .addResponseParameters(largeParam)
+            .addResponseParameters(largeParam)
+            .addResponseParameters(largeParam)
+            .addResponseParameters(largeParam)
+            .build();
+    int payloadSize = 65535 * 4 + 10;
+    StreamRecorder<Messages.StreamingOutputCallResponse> recorder = StreamRecorder.create();
+    asyncStub.streamingOutputCall(request, recorder);
+    recorder.awaitCompletion();
+    int maxWindowUpdates = (payloadSize / (65535 / 2));
+    // 1 for request & 5 window updates back to server.
+    assertTrue(6 >= ClientCountingChannel.flushCount);
+
+    // 1 flush for headers &
+    assertTrue(maxWindowUpdates + 1 >= ServerCountingChannel.flushCount);
+  }
+
+  @Test(timeout = 10000)
+  public void smallClientStreaming() throws Exception {
+    startServer(false);
+    startClient(false);
+    Messages.StreamingInputCallRequest smallPayload =
+        Messages.StreamingInputCallRequest.newBuilder()
+            .setPayload(Messages.Payload.newBuilder()
+                .setBody(ByteString.copyFrom(new byte[1])))
+            .build();
+    final List<Messages.StreamingInputCallRequest> requests = Arrays.asList(
+        smallPayload, smallPayload, smallPayload, smallPayload);
+    StreamRecorder<Messages.StreamingInputCallResponse> recorder = StreamRecorder.create();
+    StreamObserver<Messages.StreamingInputCallRequest> requestObserver =
+        asyncStub.streamingInputCall(recorder);
+    for (Messages.StreamingInputCallRequest request : requests) {
+      requestObserver.onValue(request);
+    }
+    requestObserver.onCompleted();
+    recorder.awaitCompletion();
+
+    // Maximum is 1 flush for headers, 1 for each each message & 1 for trailing empty DATA but
+    // the write queue typically coalesces to ~2.
+    assertTrue(ClientCountingChannel.flushCount <= 6);
+    // BROKEN: Should be 1 but because we don't differentiate between unary and streaming responses
+    // in generated stubs, they both use ServerCalls.asyncStreamingCall which causes us to
+    // make an unneeded call to call.request for the unary response case which triggers a flush.
+    assertTrue(ServerCountingChannel.flushCount <= 2);
+  }
+
+  private void startClient(boolean directExecutor) {
+    try {
+      NettyChannelBuilder builder = NettyChannelBuilder
+          .forAddress(new InetSocketAddress(InetAddress.getLocalHost(), serverPort))
+          .channelType(ClientCountingChannel.class)
+          .negotiationType(NegotiationType.PLAINTEXT);
+      if (directExecutor) {
+        builder.executor(MoreExecutors.newDirectExecutorService());
+      }
+      channel = builder.build();
+      blockingStub = TestServiceGrpc.newBlockingStub(channel);
+      asyncStub  = TestServiceGrpc.newStub(channel);
+      blockingStub.unaryCall(SMALL_REQUEST);
+      // Need to sleep so that SETTINGS_ACK is received and processed as it can cause a flush.
+      Thread.sleep(1000);
```

Comment:
Ouch. That means each of these tests will take at least a second. That greatly increases our testing time.

Can we instead wait for the flush count to increase due to the SETTINGS ACK? Hmm... Unfortunate all around...